feat: draw lyricsExtends, fix: lyricsSpacing

sschmidTU opened the pull request:
LyricExtends are now drawn correctly, previously weren't at all:
![image](https://user-images.githubusercontent.com/33069673/44034272-fc0c13a8-9f0c-11e8-98e5-ed10b9487866.png)

see #317, which discusses some of the changes.

lyricsSpacing was reduced, less necessary as lyricsEntries are now correctly sorted before spacing
also contains some refactors and improved comments.

No breaking changes as far as i can see, tested all demo samples with lyrics.

matt-uib commented on src/MusicalScore/Graphical/EngravingRules.ts:
```diff
@@ -122,11 +122,13 @@ export class EngravingRules {
     private wedgeLineWidth: number;
     private tupletLineWidth: number;
     private lyricUnderscoreLineWidth: number;
+    private lyricUnderscoreLineWidthVexflow: number;
```
We should not have here extra parameters for Vexflow. Please just set the paramters simply such, that it works for Vexflow. If we later need it, we might add a method setEngravingRulesForVexFlow() or make VexFlowEngravingRules or something like that.

matt-uib commented on src/MusicalScore/Graphical/EngravingRules.ts:
```diff
@@ -292,11 +294,13 @@ export class EngravingRules {
         this.wedgeLineWidth = 0.12;
         this.tupletLineWidth = 0.12;
         this.lyricUnderscoreLineWidth = 0.12;
+        this.lyricUnderscoreLineWidthVexflow = 1.3;
```
as i see it now: can there again be the factor 10 (pixels per unit) missing? Looks like it...

sschmidTU replied:
done.

matt-uib commented on src/MusicalScore/Graphical/EngravingRules.ts:
```diff
@@ -292,11 +294,13 @@ export class EngravingRules {
         this.systemRepetitionEndingLineWidth = 0.12;
         this.systemDotWidth = EngravingRules.unit / 5.0;
         this.distanceBetweenVerticalSystemLines = 0.35;
+        this.noteHeadWidth = this.distanceBetweenVerticalSystemLines * 2; // estimated
```
This should actually not be needed, as we get the bounding box from Vexflow including the note heads. So you only need to consider the right border of the BB of the end staff entry. When i look at the rendering of the extend, it seems actually slightly too long (could be the 0.7 that you add)
We can still add this amount, but let's use a different/new name for it.

sschmidTU replied:
the problem is that the lyrics extend positioning is done in MusicSheetCalculator before any VexFlow drawing, so we don't have a bounding box for the notes yet. I checked that with a debugging session as well.
I could decrease the length of the extend. They seemed fine to me, but i'll check.

matt-uib replied:
..But how can you position the line and know it's length, when there was no VexFlow layouting?

sschmidTU replied:
Good question, not sure. Maybe the Vexflow layouting is done by adding notes to voices and formatting them, but the notes aren't drawn yet. In any case, the GraphicalNote doesn't have valid BoundingBox information yet (which i checked by debugging).
But the width of the note head should pretty much always be the same, unless a rest note follows perhaps. The length of the extend also shouldn't be that critical, i think.

matt-uib commented on src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:
```diff
@@ -144,20 +144,35 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
-      for (let i: number = 0; i < measure.staffEntries.length; i++) {
-        const staffEntry: GraphicalStaffEntry = measure.staffEntries[i];
+      // sort staffEntries by position, so that we always compare neighboring staffEntries
+      const staffEntriesSorted: GraphicalStaffEntry[] = measure.staffEntries.sort(
```
Have you tried, that it doesn't work without sorting? They should be sorted actually already..

sschmidTU replied:
Yes, i checked that with a debugging session, they can be unsorted. For example in An die Musik, for "war-mer Lieb' ent-zun-den", i had the StaffEntries with lyricEntries ["war","mer","ent","Lieb'"] and so on.
I also thought they should be sorted, because that's done somewhere in the code, i'd have to look where that is. But i'm pretty sure the entries weren't sorted, and that's why i got much better spacing with much less minDistance (0.3 -> 0.1) after sorting.

matt-uib replied:
interesting... maybe that's the reason for other problems. Let's keep an eye on it..

sschmidTU commented:
43efa1a63c4cac715dfef3188cf28ae5f094631:
* Removed the Vexflow engraving rule.
* Lyrics spacing is now done for each verse.
* Added the Austrian National Anthem to the demo samples (it's public domain).
It's interesting for spacing, i will add an issue for spacing the last lyric in a measure ("zukunftsreich"), maybe shouldn't go into this PR.

54af65a906e171b4cd81026162971afdc7a5f50c:
* sorting staffEntries/lyricEntries removed from lyrics spacing, not necessary after:
* fix double lyrics width calculation
(accidentally called in both MusicSheetCalculator and VexFlowMusicSheetCalculator)
* shortened lyric extend
not extended by noteHeadWidth anymore. Length is still fine, we have more room for spacing, and sheet music editions do all kinds of extend lengths anyways, mostly to the right edge of the next note head, sometimes to the middle of it.

@matt-uib i think this PR is ready now, i fixed all the issues you mentioned. We could decide not to include the Austrian National Anthem (public domain) in this PR, but it already looks pretty nice, except for the one interesting little spacing issue.
